velox/common/compression/v2/Lz4Compression.cpp

```diff
@@ -0,0 +1,529 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <glog/logging.h>
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/compression/v2/Lz4Compression.h"
+
+namespace facebook::velox::common {
+
+namespace {
+
+void lz4Error(LZ4F_errorCode_t errorCode, const char* prefixMessage) {
+  VELOX_FAIL(prefixMessage, LZ4F_getErrorName(errorCode));
+}
+
+LZ4F_preferences_t defaultPreferences() {
+  LZ4F_preferences_t prefs;
+  memset(&prefs, 0, sizeof(prefs));
+  return prefs;
+}
+
+LZ4F_preferences_t defaultPreferences(int compressionLevel) {
+  LZ4F_preferences_t prefs = defaultPreferences();
+  prefs.compressionLevel = compressionLevel;
+  return prefs;
+}
+} // namespace
+
+class LZ4Compressor : public Compressor {
+ public:
+  explicit LZ4Compressor(int32_t compressionLevel);
+
+  ~LZ4Compressor() override;
+
+  void init();
+
+  CompressResult compress(
+      uint64_t inputLength,
+      const uint8_t* input,
+      uint64_t outputLength,
+      uint8_t* output) override;
+
+  FlushResult flush(uint64_t outputLength, uint8_t* output) override;
+
+  EndResult end(uint64_t outputLength, uint8_t* output) override;
+
+ protected:
+  void
+  compressBegin(uint8_t* output, size_t& outputLen, uint64_t& bytesWritten);
+
+  int compressionLevel_;
+  LZ4F_compressionContext_t ctx_{nullptr};
+  LZ4F_preferences_t prefs_;
+  bool firstTime_;
+};
+
+class LZ4Decompressor : public Decompressor {
+ public:
+  LZ4Decompressor() {}
+
+  ~LZ4Decompressor() override {
+    if (ctx_ != nullptr) {
+      LZ4F_freeDecompressionContext(ctx_);
+    }
+  }
+
+  void init();
+
+  void reset() override;
+
+  DecompressResult decompress(
+      uint64_t inputLength,
+      const uint8_t* input,
+      uint64_t outputLength,
+      uint8_t* output) override;
+
+  bool isFinished() override;
+
+ protected:
+  LZ4F_decompressionContext_t ctx_ = nullptr;
+  bool finished_;
+};
+
+LZ4Compressor::LZ4Compressor(int32_t compressionLevel)
+    : compressionLevel_(compressionLevel) {}
+
+LZ4Compressor::~LZ4Compressor() {
+  if (ctx_ != nullptr) {
+    LZ4F_freeCompressionContext(ctx_);
+  }
+}
+
+void LZ4Compressor::init() {
+  LZ4F_errorCode_t ret;
+  prefs_ = defaultPreferences(compressionLevel_);
+  firstTime_ = true;
+
+  ret = LZ4F_createCompressionContext(&ctx_, LZ4F_VERSION);
+  if (LZ4F_isError(ret)) {
+    lz4Error(ret, "LZ4 init failed: ");
```

Comment:
Any missing content for the error message?

Reply:
`lz4Error` will try to expand the error message for the return code. Switched the param order to make it clear for the reader.